test/performance/odp_packet_gen.c

```diff
@@ -0,0 +1,1232 @@
+/* Copyright (c) 2020, Nokia
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+
+#define MAX_PKTIOS        32
+#define MAX_PKTIO_NAME    255
+#define RX_THREAD         1
+#define TX_THREAD         2
+
+/* Minimum number of packets to receive in CI test */
+#define MIN_RX_PACKETS_CI 800
+
+typedef struct test_options_t {
+	uint64_t gap_nsec;
+	uint64_t quit;
+	uint32_t num_rx;
+	uint32_t num_tx;
+	uint32_t num_cpu;
+	uint32_t num_pktio;
+	uint32_t num_pkt;
+	uint32_t pkt_len;
+	uint32_t burst_size;
+	uint32_t bursts;
+	uint32_t ipv4_src;
+	uint32_t ipv4_dst;
+	uint16_t udp_src;
+	uint16_t udp_dst;
+
+	struct {
+		uint32_t udp_src;
+		uint32_t udp_dst;
+	} c_mode;
+
+	char     pktio_name[MAX_PKTIOS][MAX_PKTIO_NAME + 1];
+	char     ipv4_src_s[24];
+	char     ipv4_dst_s[24];
+
+} test_options_t;
+
+typedef struct thread_arg_t {
+	void *global;
+	int tx_thr;
+
+	/* pktout queue per pktio interface (per thread) */
+	odp_pktout_queue_t pktout[MAX_PKTIOS];
+
+} thread_arg_t;
+
+typedef struct ODP_ALIGNED_CACHE thread_stat_t {
+	uint64_t time_nsec;
+	uint64_t rx_timeouts;
+	uint64_t rx_packets;
+	uint64_t rx_bytes;
+
+	uint64_t tx_timeouts;
+	uint64_t tx_packets;
+	uint64_t tx_drops;
+
+	int      thread_type;
+
+} thread_stat_t;
+
+typedef struct test_global_t {
+	test_options_t test_options;
+	odp_atomic_u32_t exit_test;
+	odp_barrier_t barrier;
+	odp_cpumask_t cpumask;
+	odp_pool_t pool;
+	odph_thread_t thread_tbl[ODP_THREAD_COUNT_MAX];
+	thread_stat_t stat[ODP_THREAD_COUNT_MAX];
+	thread_arg_t thread_arg[ODP_THREAD_COUNT_MAX];
+
+	struct {
+		odph_ethaddr_t eth_src;
+		odph_ethaddr_t eth_dst;
+		odp_pktio_t pktio;
+		odp_pktout_queue_t pktout[ODP_THREAD_COUNT_MAX];
+		int started;
+
+	} pktio[MAX_PKTIOS];
+
+} test_global_t;
+
+test_global_t test_global;
+
+static void print_usage(void)
+{
+	printf("\n"
+	       "ODP packet generator\n"
+	       "\n"
+	       "Usage: odp_packet_gen [options]\n"
+	       "\n"
+	       "  -i, --interface <name>  Packet IO interfaces (comma-separated, no spaces)\n"
```

Comment:
Separate mandatory and optional options. Perhaps an example could be useful also.